include/swift/AST/SILGenRequests.h

```diff
@@ -96,6 +97,30 @@ void simple_display(llvm::raw_ostream &out, const SILGenDescriptor &d);
 
 SourceLoc extractNearestSourceLoc(const SILGenDescriptor &desc);
 
+class SILGenSourceFileRequest :
+    public SimpleRequest<SILGenSourceFileRequest,
+                         bool(Lowering::SILGenModule *, SourceFile *),
+                         CacheKind::Uncached> {
+public:
+  using SimpleRequest::SimpleRequest;
+
+private:
+  friend SimpleRequest;
+
+  // Evaluation.
+  llvm::Expected<bool>
+  evaluate(Evaluator &evaluator,
+           Lowering::SILGenModule *SGM, SourceFile *SF) const;
```

Comment:
If you split this up into two requests, you can make the SGM completely internal to the request itself:

- SILGenSourceFileRequest would only be used in non-WMO builds, and create a new SGM for the source file
- SILGenWholeModuleRequest would in turn be used in WMO builds, and it would take an AST ModuleDecl and create a new SGM and emit all source files.

Since WMO builds shouldn't be generating incremental dependency info, it shouldn't actually matter if the "top level request" in WMO mode doesn't have a source location.